chainerrl/envs/vec_env.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+from builtins import *  # NOQA
+from future import standard_library
+standard_library.install_aliases()  # NOQA
+
+from multiprocessing import Pipe
+from multiprocessing import Process
+
+from cached_property import cached_property
+import numpy as np
+
+from chainerrl import env
+
+
+def worker(remote, env_fn_wrapper):
+    env = env_fn_wrapper
+    while True:
+        cmd, data = remote.recv()
+        if cmd == 'step':
+            ob, reward, done, info = env.step(data)
+            if done:
+                ob = env.reset()
+            remote.send((ob, reward, done, info))
+        elif cmd == 'reset':
+            ob = env.reset()
+            remote.send(ob)
+        elif cmd == 'close':
+            remote.close()
+            break
+        elif cmd == 'get_spaces':
+            remote.send((env.action_space, env.observation_space))
+        elif cmd == 'spec':
+            remote.send(env.spec)
+        elif cmd == 'seed':
+            remote.send(env.seed(data))
+        else:
+            raise NotImplementedError
+
+
+class VectorEnv(env.Env):
+
+    def __init__(self, env_fns):
+        """envs: list of gym environments to run in subprocesses
+
+        """
+        nenvs = len(env_fns)
+        self.remotes, self.work_remotes = zip(*[Pipe() for _ in range(nenvs)])
+        self.ps = \
+            [Process(target=worker, args=(work_remote, env_fn))
+             for (work_remote, env_fn) in zip(self.work_remotes, env_fns)]
+        for p in self.ps:
+            p.start()
+
+        self.remotes[0].send(('get_spaces', None))
+        self.action_space, self.observation_space = self.remotes[0].recv()
+
+    @cached_property
+    def spec(self):
+        self.remotes[0].send(('spec', None))
+        spec = self.remotes[0].recv()
+        return spec
+
+    def step(self, actions):
+        for remote, action in zip(self.remotes, actions):
+            remote.send(('step', action))
+        results = [remote.recv() for remote in self.remotes]
+        obs, rews, dones, infos = zip(*results)
+        return np.stack(obs), np.stack(rews), np.stack(dones), infos
+
+    def reset(self):
+        for remote in self.remotes:
+            remote.send(('reset', None))
+        return np.stack([remote.recv() for remote in self.remotes])
+
+    def close(self):
```

Comment:
If `close` is not called, the program cannot exit due to remaining processes. This is cumbersome. I think we should do the same procedure in `VectorEnv.__del__` instead so that users do not need to care about it.